internal/controller/csiaddons/persistentvolumeclaim_controller.go

```diff
@@ -241,48 +169,87 @@ func (r *PersistentVolumeClaimReconciler) checkDriverSupportReclaimsSpace(logger
-//   - If schedule is present in namespace annotations, check for reclaimSpace
-//     support by the driver.
-//   - If driver supports reclaimSpace, use the schedule from namespace.
-//   - If driver does not support reclaimSpace, return ErrScheduleNotFound.
-//     Depending on requeue value, it will throw ErrorConnNotFoundRequeueNeeded.
+//   - Check if the schedule is present in the storageclass annotations. If yes,
+//     use that.
 func (r *PersistentVolumeClaimReconciler) determineScheduleAndRequeue(
 	ctx context.Context,
 	logger *logr.Logger,
 	pvc *corev1.PersistentVolumeClaim,
 	driverName string,
+	annotationKey string,
 ) (string, error) {
 	annotations := pvc.GetAnnotations()
-	schedule, scheduleFound := getScheduleFromAnnotation(logger, annotations)
+	schedule, scheduleFound := getScheduleFromAnnotation(annotationKey, logger, annotations)
 	if scheduleFound {
 		return schedule, nil
 	}
 
 	// check for namespace schedule annotation.
-	// We cannot have a generic solution for all CSI drivers to get the driver
-	// name from PV and check if driver supports space reclamation or not and
-	// requeue the request if the driver is not registered in the connection
-	// pool. This can put the controller in a requeue loop. Hence we are
-	// reading the driver name from the namespace annotation and checking if
-	// the driver is registered in the connection pool and if not we are not
-	// requeuing the request.
 	ns := &corev1.Namespace{}
 	err := r.Client.Get(ctx, types.NamespacedName{Name: pvc.Namespace}, ns)
 	if err != nil {
 		logger.Error(err, "Failed to get Namespace", "Namespace", pvc.Namespace)
 		return "", err
 	}
-	schedule, scheduleFound = getScheduleFromAnnotation(logger, ns.Annotations)
+	schedule, scheduleFound = getScheduleFromAnnotation(annotationKey, logger, ns.Annotations)
 
 	// If the schedule is found, check whether driver supports the
 	// space reclamation using annotation on namespace and registered driver
 	// capability for decision on requeue.
 	if scheduleFound {
+		// If the check is for key rotation
+		// check if the driver supports the operation
+		if annotationKey == krcJobScheduleTimeAnnotation {
+			requeue, keyRotationSupported := r.checkDriverSupportsEncryptionKeyRotate(logger, ns.Annotations, driverName)
+			if keyRotationSupported {
+				return schedule, nil
+			}
+			if requeue {
+				return "", ErrConnNotFoundRequeueNeeded
+			}
+		}
+
+		// We cannot have a generic solution for all CSI drivers to get the driver
+		// name from PV and check if driver supports space reclamation or not and
+		// requeue the request if the driver is not registered in the connection
+		// pool. This can put the controller in a requeue loop. Hence we are
+		// reading the driver name from the namespace annotation and checking if
+		// the driver is registered in the connection pool and if not we are not
+		// requeuing the request.
+		// Depending on requeue value, it will return ErrorConnNotFoundRequeueNeeded.
```

Comment:
move this comment to top of above if check and move reclaimspace check into `else if annotationKey == reclaimspaceScheduleTimeAnnotation`